File: verifier-ray/codegen/rowlimit.go
package codegen

import (
	"fmt"

	"github.com/LFDT-Lineth/lineth-monorepo/prover-ray/wiop"
	"github.com/LFDT-Lineth/lineth-monorepo/prover-ray/wiop/compilers/lookuptologderivsum"
)

// RowLimitSystem is the compiled metadata for every lookup row-limit check in a
// wiop.System, in the form the Zig rowlimit sub-verifier consumes.
//
// prover-ray's lookuptologderivsum compiler bin-packs lookups sharing an
// includings table into subgroups, each of which shares a multiplicity
// column M and therefore drains its own MaxLookupRows budget independently
// (see lookuptologderivsum.RowLimitVerifierAction). Compile-time bin-packing
// keeps every subgroup under budget using static (maximum) module heights, but
// a dynamic module's RUNTIME size is prover-declared and PCS does not bind it
// to anything — nothing else in the transcript catches a prover that declares
// a far larger size than it compiled against. This sub-verifier re-checks the
// exact per-run heights for the concrete proof, exactly mirroring the runtime
// check prover-ray's own verifier performs.
type RowLimitSystem struct {
	SourceName string
	Checks     []RowLimitCheck
}

// RowLimitCheck is one subgroup's row-limit check: the included and
// includings module partitioning (one ModuleSize per fragment, mirroring
// RowLimitVerifierAction.IncludedModules/IncludingsModules one-to-one, so a
// repeated module appears once per fragment exactly as prover-ray's own sum
// does) and the shared per-side budget.
type RowLimitCheck struct {
	IncludedModules   []ModuleSize
	IncludingsModules []ModuleSize
	Limit             uint64
}

// BuildRowLimitSystem extracts every lookuptologderivsum.RowLimitVerifierAction
// registered on sys and records the module partitioning each one enforces.
// Checks are collected in round/registration order so the output is
// deterministic.
func BuildRowLimitSystem(sys *wiop.System) (RowLimitSystem, error) {
	out := RowLimitSystem{SourceName: sys.Context.Path()}
	dynamicIndices := DynamicModuleIndex(sys)

	moduleSizes := func(modules []*wiop.Module) ([]ModuleSize, error) {
		sizes := make([]ModuleSize, len(modules))
		for i, m := range modules {
			if m.IsDynamic() {
				idx, ok := dynamicIndices[m]
				if !ok {
					return nil, fmt.Errorf("codegen: dynamic module %q not found in sys.Modules order", m.Context.Path())
				}
				sizes[i] = ModuleSize{Dynamic: true, DynamicIndex: idx}
			} else {
				sizes[i] = ModuleSize{StaticSize: m.Size()}
			}
		}
		return sizes, nil
	}

	for _, round := range sys.Rounds {
		for _, action := range round.VerifierActions {
			rl, ok := action.(*lookuptologderivsum.RowLimitVerifierAction)
			if !ok {
				continue
			}

			includedModules, err := moduleSizes(rl.IncludedModules)
			if err != nil {
				return RowLimitSystem{}, err
			}
			includingsModules, err := moduleSizes(rl.IncludingsModules)
			if err != nil {
				return RowLimitSystem{}, err
			}

			out.Checks = append(out.Checks, RowLimitCheck{
				IncludedModules:   includedModules,
				IncludingsModules: includingsModules,
				Limit:             wiop.MaxLookupRows,
			})
		}
	}

	return out, nil
}

File: verifier-ray/src/query/rowlimit.zig
pub const Error = error{
    MissingDynamicModuleSize,
    RowLimitExceeded,
};

/// The runtime size of one module contributing to a row-limit check's side sum.
/// Mirrors vanishing.ModuleSize; kept as its own copy rather than shared,
/// following this codebase's convention of each sub-verifier owning its data
/// shapes independently (see e.g. pcs.SizeSource, which is shaped differently
/// again for its own needs).
pub const ModuleSize = union(enum) {
    static: usize,
    dynamic: usize,
};

/// One subgroup's row-limit check: prover-ray's lookuptologderivsum compiler
/// bin-packs lookups sharing an includings table into subgroups that share a
/// multiplicity column M, so each subgroup drains its own row budget
/// independently. included_modules/includings_modules list one ModuleSize per
/// fragment on each side (a repeated module appears once per fragment,
/// deliberately double-counting it — each fragment is its own pass over that
/// module's rows).
pub const Check = struct {
    included_modules: []const ModuleSize,
    includings_modules: []const ModuleSize,
    limit: u64,
};

pub const System = struct {
    checks: []const Check = &.{},
};

/// Sums the runtime row count of every module in `modules`, resolving dynamic
/// sizes from `module_sizes` (the proof-supplied per-dynamic-module sizes, in
/// the same order prover-ray's Runtime.AdvanceRound absorbs them).
fn sumRows(modules: []const ModuleSize, module_sizes: []const usize) Error!u64 {
    var sum: u64 = 0;
    for (modules) |m| {
        sum += switch (m) {
            .static => |n| n,
            .dynamic => |idx| blk: {
                if (idx >= module_sizes.len) return error.MissingDynamicModuleSize;
                break :blk module_sizes[idx];
            },
        };
    }
    return sum;
}

/// Verifies every row-limit check in `system`. This is the verifier-side
/// counterpart to prover-ray's lookuptologderivsum.RowLimitVerifierAction.Check:
/// a dynamic module's runtime size is prover-declared and unbound by anything
/// else in the transcript (PCS commits to the module's data, not its claimed
/// size), so this check is what stops a prover from claiming a far larger
/// dynamic size than it compiled against — one whose accumulators would
/// overflow the small field the reduced lookup constraints run in.
pub fn verify(comptime system: System, module_sizes: []const usize) Error!void {
    inline for (system.checks) |check| {
        const included_rows = try sumRows(check.included_modules, module_sizes);
        if (included_rows >= check.limit) return error.RowLimitExceeded;

        const includings_rows = try sumRows(check.includings_modules, module_sizes);
        if (includings_rows >= check.limit) return error.RowLimitExceeded;
    }
}
